### PR TITLE
Pull Request for 315

### PR DESCRIPTION
**Pull request by tn3rb**
Not sure what you have and haven't seen in this branch, and for whatever reason GH is showing a bunch of commits for CB ticket 11400.
You can likely skip those, but will want to look at those for CB ticket 11426 as well as those for #315 
A couple of the commits here are also for #316

### CODE REVIEW

**Comment by nerrad** [comment]
@tn3rb It's showing the commits from the other branch because you based this pull off of master.  I'll redo the pull and copy your comments into a new pull request based off of the branch it gets merged back into.

**Comment by nerrad** [comment]
Actually I was able to edit this pull and change the base that way.  So should be good now (aside from the obvious conflicts.  Just means it has to get a merge in from the base branch before you can merge back via github.

**Comment by nerrad** [comment]
connect with #316

File: assets/src/blocks/widgets/event-attendees.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+const {__}                = wp.i18n; // Import __() from wp.i18n
```
Comment (nerrad) [inline assets/src/blocks/widgets/event-attendees.js]: This should be `import { __ } from '@wordpress/i18n'`  because we import from the package now.

File: assets/src/blocks/widgets/event-attendees.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      save: function(props) {
+        return (
+            <div className={props.className}>
+              <p>Next Upcoming Event Datetime:</p>
+            </div>
+        );
+      },
+    },
+);
+
```
Comment (nerrad) [inline assets/src/blocks/widgets/event-attendees.js]: Please run `npm run lint` in this branch and/or set up your PhpStorm via [this doc](https://github.com/eventespresso/event-espresso-core/blob/master/docs/AA--Javascript-in-EE/javscript-coding-standards.md) for the javascript styles we're following.

File: core/domain/entities/editor/blocks/widgets/EventAttendees.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+     * including setting the block type and supported post types
+     *
+     * @return void
+     */
+    public function initialize()
+    {
+        $this->setEditorBlockType(EventAttendees::BLOCK_TYPE);
+        $this->setSupportedPostTypes(array('espresso_events', 'post', 'page'));
+        $this->setAttributes(array());
+        $this->setAssetRegisterFqcn('EventEspresso\core\domain\entities\editor\blocks\CoreBlocksAssetRegister');
```
Comment (nerrad) [inline core/domain/entities/editor/blocks/widgets/EventAttendees.php]: Might be unneeded if the block is instantiated with an instance of `AssetRegister`?

File: core/domain/services/admin/ExitModal.php [hunk 4]
```diff
@@ -2,6 +2,7 @@
 namespace EventEspresso\core\domain\services\admin;
 
 use EventEspresso\core\services\assets\Registry;
+use const FILTER_VALIDATE_BOOLEAN;
```
Comment (nerrad) [inline core/domain/services/admin/ExitModal.php]: This will cause fatals  for PHP versions < 5.6 ... it's been removed multiple times, I wonder how it crept back in :(

File: core/domain/values/session/SessionLifespan.php [hunk 5]
```diff
@@ -104,11 +104,11 @@ public function inHoursMinutesSeconds($separator = ':')
      *
      * @param bool $utc If true, displays expiration in UTC
      *                  If false, displays expiration in local time
-     * @return int
+     * @return float
      */
     public function expiration($utc = true)
     {
-        return (int) current_time('timestamp', $utc) - $this->lifespan;
+        return current_time('timestamp', $utc) - $this->lifespan;
```
Comment (nerrad) [inline core/domain/values/session/SessionLifespan.php]: did you merge from 11400 or directly from master?  This looks like a revert of changes made on master.  As mentioned yesterday I already took care of a bunch of conflicts that happened when merging master into 11400 so if you merged from 11400 then I might have resolved this conflict wrong. Yet, I'm pretty sure I resolved it to use `int` (from master) so the fact that I noticed this  here is kind of odd.

File: core/services/assets/Registry.php [hunk 6]
```diff
@@ -634,8 +636,10 @@ private function removeAlreadyRegisteredDataForScriptHandle($script_handle)
         if (isset($this->script_handles_with_data[$script_handle])) {
             global $wp_scripts;
             if ($wp_scripts->get_data($script_handle, 'data')) {
-                unset($wp_scripts->registered[$script_handle]->extra['data']);
-                unset($this->script_handles_with_data[$script_handle]);
+                unset(
+                    $wp_scripts->registered[$script_handle]->extra['data'],
+                    $this->script_handles_with_data[$script_handle]
+                );
```
Comment (nerrad) [inline core/services/assets/Registry.php]: k this is a bit alarming because it seems to indicate that the branch this is in is out of date. The actual code in this method _currently_ within master and `FET/11400/block-manager` is:

```
private function removeAlreadyRegisteredDataForScriptHandle($script_handle)
    {
        if (isset($this->script_handles_with_data[$script_handle])) {
            global $wp_scripts;
            $unset_handle = false;
            if ($wp_scripts->get_data($script_handle, 'data')) {
                unset($wp_scripts->registered[$script_handle]->extra['data']);
                $unset_handle = true;
            }
            //deal with inline_scripts
            if ($wp_scripts->get_data($script_handle, 'before')) {
                unset($wp_scripts->registered[$script_handle]->extra['before']);
                $unset_handle = true;
            }
            if ($wp_scripts->get_data($script_handle, 'after')) {
                unset($wp_scripts->registered[$script_handle]->extra['after']);
            }
            if ($unset_handle) {
                unset($this->script_handles_with_data[$script_handle]);
            }
        }
    }
```

So we've got a problem if its not this in the branch you are editing.  There will be some nasty conflicts to try to resolve because of the additional changes you've made in this class in this branch (and possible reverts of work could happen too)

File: core/services/collections/Collection.php [hunk 7]
```diff
@@ -4,6 +4,7 @@
 use  EventEspresso\core\exceptions\InvalidEntityException;
 use  EventEspresso\core\exceptions\InvalidInterfaceException;
 use LimitIterator;
+use SplObjectStorage;
```
Comment (nerrad) [inline core/services/collections/Collection.php]: since gutenberg branches are going to live outside of master for a bit.  I think it'd be wiser if you make these kind of updates in separate branches off master do a pull request, and merge them in.  Otherwise you increase the risk of conflicts when/if the same kind of cleanup happens in other branches.

File: core/services/collections/Collection.php [hunk 8]
```diff
@@ -38,6 +47,37 @@ class Collection extends \SplObjectStorage implements CollectionInterface {
...
+
+
+	 /**
+      * creates a very readable unique 9 character identifier like:  CF2-532-DAC
+      * and appends it to the non-qualified class name, ex: ThingCollection-CF2-532-DAC
+      *
+	  * @return void
+	  */
+	 protected function setCollectionIdentifier() {
+         // hash a few collection details
```
Comment (nerrad) [inline core/services/collections/Collection.php]: I'm curious to know the reasoning for this beyond just creating a human friendly identifier.  In other words, **why** are you creating a human friendly identifier?

Also, again, this is something that I think should be done in a separate branch as an atomic commit rather than lumped in with a potentially long living branch.

File: src/assets/src/blocks/query-panel/index.js [hunk 9]
```diff
@@ -0,0 +1,127 @@
+/**
```
Comment (nerrad) [inline src/assets/src/blocks/query-panel/index.js]: all these files within `/src/assets/src` are in the wrong path.  I also don't think we should bring these into these EE core branches from `eea-gutenberg` until we actually implement them.

This is also is a possible indicator that you've merged from a different branch or have NOT merged from the `FET/Gutenberg/11400/block-manager` branch that I took care of updating from master yesterday because in that branch I deleted all these files (or maybe you saw they were deleted and added them back?)

**Comment by tn3rb** [comment]
> It's showing the commits from the other branch because you based this pull off of master.

No I did not base this pull off of master. I tried initiating the pull request from my Git client (GitKraken ), which has to be done from the specific branch, so in other words, this was initiated from the `FET/Gutenberg/11426/event-attendees-block` branch. It looks to me like it just dumped in ALL of the commits since the branch had been cut. If my Git client had set the base to master, then my apologies, as that was not my intent. To avoid any future issues, I will not initiate any more PRs from GitKraken since it appears to do bad things.

> So when this work is merged, will it also complete #315 and #316?

No. This PR is **only** connected to #315. In order to get things working with the new build process changes you introduced, I needed to make some changes to the underlying block managers as well as the parent block.

> Since this branch is based off of FET/Gutenberg/11400/block-manager there's something awry and maybe its just because it hasn't been updated from that branch. Regardless, there will be a lot of conflicts and potential unnoticed reverts of work if this branch is merged back into that branch. Something has gone south here and needs fixed.
> This may be related to the branches not being in sync, but I'm concerned about the number of commits/changes in non gutenberg specific files that are showing up in this pull. These kinds of changes should be done in their own branches in more atomic changesets to reduce changes of unexpected regressions and or conflicts.

This branch just needed to be updated again from 11400. I've done that now and it's fine. You were looking at old cruft. Sry for forgetting to update before pushing. I did this later in the evening last night when I was tired (granted, I should have updated the branch prior to working on it)

> The AssetRegister stuff looks good (aside from the comments I made in a few places). But imo it would have been better had you implemented in it's OWN branch off of FET/Gutenberg/11400/block-manager and done a pull with just those changes and the files they change. The event-attendees-block branch should just be for doing an event-attendees-block. The purpose for the base branch was so that any Framework stuff is always kept up to date in that branch and the block branches are just based off of that branch and not making changes to framework stuff in their branches. Otherwise we will have conflict hell. We can talk more about this in slack if I'm not clear.



Before I could build any blocks, I needed to build the framework for loading blocks and their corresponding assets, etc.
I did that in 11400 and then started to work on 11426 to build the event attendees block that will replace the shortcode of the same name.
Then the new build process for assets was introduced and the above work stopped functioning. Which is fine.
But in order to build blocks such as the event attendees block, I need to fix the system for loading blocks and their corresponding assets, etc.,
and to do that, I need at least one working block so that I can determine if the system for loading blocks is fixed or not.

You may have noticed that the actual JS for the event attendees block does not incorporate any of the changes regarding the new build process.
That's because I haven't touched it yet because I needed to get the underlying system working first. Once the framework stuff is working, I can merge this branch into 11400, and then either start a new branch solely in GH for finishing the event attendees block, or just continue working on this one (I'll most likely cut a new branch). 

I understand your point about "not making changes to framework stuff" in other branches, but if framework changes are required to get Feature-X working, don't you think it might be difficult to do so in a branch that doesn't have Feature-X in it? How are you going to know when it is actually working or not? Like imagine trying to build a new framework for loading EE admin pages in a new branch that didn't have any actual admin page classes in it!!! Kinda tricky eh? So  I could have created a new branch from 11400 for doing this work, but I still would have had to include the event attendees block in order to actually know it was working, so... what's the difference? Either way you end up with framework changes **and a block** in your branch. It was easier and faster to just use the existing branch.

Anyways, I should have been more explicit in clarifying that this pull request was for #315 and that it needs to be completed and merged into 11400 before any further work on individual blocks can progress. This PR **is** connected to issue #315 but that's not really that obvious until you get to the very bottom of the screen and see where that detail is documented.

> Also with regards to branch naming. I've been following the schema of FET/Gutenberg/11400/{issue/ticketnum}/description so that its clear from the branch name that its based off of the 11400 branch. I think we need to try to have a way to be clear about branch base.

well how about we create a new Gutenberg base branch that's just named `FET/Gutenberg` or something similar? I certainly don't see the need of keeping 11400 in the name since that is a reference to the old Codebase ticket. We also don't want to have multiple base branches from which we are creating new branches, ie: we don't want to have `FET/Gutenberg/11426/{issue}/description` or `FET/Gutenberg/11500/{issue}/description` branches right? So maybe let's just drop 11400 from the branch name. 

Since we now have the option to create Projects and Epics, maybe we can use those as part of the naming schema. For example, I create a Project named "Gutenberg Integration" as a test just to see what was involved with that. So maybe we could move all of the Gutenberg related issues into that project, and then instead of using `FET` or `BUG` for the base of our branch names, we use the Project name (or an abbreviation of it) like `Gutenberg/{issue}/description` or `GTB/{issue}/description` ? So for example, instead of `FET/Gutenberg/11426/event-attendees-block` that branch would be named `Gutenberg/316/event-attendees-block` or `GTB/316/event-attendees-block`.

So branches starting with  `FET` or `BUG` indicate that they are cut from master, and any branch name starting with something else indicates a larger feature Project that will eventually get merged into master

Thoughts?

**Comment by nerrad** [comment]
> That's because I haven't touched it yet because I needed to get the underlying system working first. Once the framework stuff is working, I can merge this branch into 11400, and then either start a new branch solely in GH for finishing the event attendees block, or just continue working on this one (I'll most likely cut a new branch).

> I understand your point about "not making changes to framework stuff" in other branches, but if framework changes are required to get Feature-X working, don't you think it might be difficult to do so in a branch that doesn't have Feature-X in it? How are you going to know when it is actually working or not? Like imagine trying to build a new framework for loading EE admin pages in a new branch that didn't have any actual admin page classes in it!!! Kinda tricky eh? So I could have created a new branch from 11400 for doing this work, but I still would have had to include the event attendees block in order to actually know it was working, so... what's the difference? Either way you end up with framework changes and a block in your branch. It was easier and faster to just use the existing branch.

I felt I had to make the point I did because it wasn't clear that's what you were already planning on doing because of all the stuff that still in the branch and the name of the branch etc.  So without that clarity I felt I had to make that point.  I totally dig what you're saying and agree sometimes you need a bare minimum of implementation to test functionality.  I'm satisfied that you plan on merging back into the base branch as soon as you know it will work.

> well how about we create a new Gutenberg base branch that's just named FET/Gutenberg or something similar?

Yup me like.

> Since we now have the option to create Projects and Epics, maybe we can use those as part of the naming schema. For example, I create a Project named "Gutenberg Integration" as a test just to see what was involved with that. So maybe we could move all of the Gutenberg related issues into that project, and then instead of using FET or BUG for the base of our branch names, we use the Project name (or an abbreviation of it) like Gutenberg/{issue}/description or GTB/{issue}/description ? So for example, instead of FET/Gutenberg/11426/event-attendees-block that branch would be named Gutenberg/316/event-attendees-block or GTB/316/event-attendees-block.

Ya I'm chill with doing that.  However, I imagine that there will be some cases where we simply create a pull request devoid of any issue number.  Because of how well pull requests work I don't think issue num has to be a requirement in the branch name.  So our pattern for branch names could be something like this:

```
{PROJECT_NAME/EPIC_NAME}|BUG|FET/[{issue-num}/]{concise-description}
```

_Base_  branches for a project should simply be `{PROJECT_NAME|EPIC_NAME}/base`

...With all the benefits  you mentioned.

Thoughts?

File: assets/src/blocks/widgets/event-attendees.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
+const {__}                = wp.i18n; // Import __() from wp.i18n
+const {registerBlockType} = wp.blocks; // Import registerBlockType() from wp.blocks
+
+registerBlockType(
+    'event-espresso/widgets-event-attendees',
+    {
+      title:       __('Event Attendees'),
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/src/blocks/widgets/event-attendees.js]: will do in a PR for #316

File: assets/src/blocks/widgets/event-attendees.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+      description: 'Displays a list of people that have registered for the specified event',
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/src/blocks/widgets/event-attendees.js]: will do in a PR for #316

File: assets/src/blocks/widgets/event-attendees.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+          source: 'children',
+          selector:
+                  'p',
+          type:
+                  'string',
+        },
+      },
+      // The "edit" property must be a valid function.
+      edit:        function(props) {
+        console.log(props);
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/src/blocks/widgets/event-attendees.js]: will fix in a PR for #316

File: assets/src/blocks/widgets/event-attendees.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+        return (
+            <div className={props.className}>
+              <p>Event Editor</p>
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/src/blocks/widgets/event-attendees.js]: ya just ignore plz

File: assets/src/blocks/widgets/event-attendees.js [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+            </div>
+        );
+      },
+
+      // The "save" property must be specified and must be a valid function.
+      //   <p>props.attributes.id</p>
+      save: function(props) {
+        return (
+            <div className={props.className}>
+              <p>Next Upcoming Event Datetime:</p>
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/src/blocks/widgets/event-attendees.js]: ya just ignore plz

File: .github/label-sync.yml [hunk 15]
```diff
@@ -30,6 +30,20 @@
...
+    - name: "status:next"
+      color: efefef
+    - name: "status:needs-testing"
+      color: efefef
+    - name: "status:merge"
+      color: efefef
+    - name: "status:review"
+      color: efefef
+    - name: "status:feedback-needed"
+      color: efefef
```
Comment (nerrad) [inline .github/label-sync.yml]: it's things like this that make me think there was a merge to this branch directly from master instead of to 11400 and then this branch.  Currently in 11400 I'm NOT seeing these changes from master.

Comment (nerrad) [reply .github/label-sync.yml]: Brent and i chatted about this in slack and got it resolved, 11400 just hadn't been pushed yet.

File: core/services/editor/EditorBlockAssetRegister.php [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+
+/**
+ * Class EditorBlockAssetRegister
+ * Abstract parent for classes that handle asset registration for one or more editor blocks that share the same assets
+ *
+ * @package EventEspresso\core\services\editor
+ * @author  Brent Christensen
+ * @since   $VID:$
+ */
+abstract class EditorBlockAssetRegister implements AssetRegisterInterface
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply core/services/editor/EditorBlockAssetRegister.php]: ya I initially didn't like the idea of **just** using "block" to reference these things because it just seems way too generic, but that said, we are using namespacing which pretty much negates that issue, and "block" is what WP uses, so ya... I'll change all usages of "EditorBlock" to just "Block"

File: core/services/editor/EditorBlockAssetRegister.php [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+        DomainInterface $domain,
+        Registry $registry
+    ) {
+        $this->script_handle       = $script_handle;
+        $this->script_dependencies = $script_dependencies;
+        $this->style_handle        = $style_handle;
+        $this->style_dependencies  = $style_dependencies;
+        $this->domain              = $domain;
+        $this->registry            = $registry;
+    }
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply core/services/editor/EditorBlockAssetRegister.php]: **Absolutely No**!!!

That is a bad idea that should be avoided like the plague, so please don't do this kind of thing anywhere, and I'll explain why.

A quick example that hopefully makes it instantly clear is "unit testing". 

The longer answer is that a class should directly declares its dependencies by type hinting for them in it's constructor, which allows other classes to know what it requires in order to operate. Obtaining dependencies via other objects, hides a class's dependencies and is a form of "service location", which is discouraged for many reasons (you can read more about that if you want to by Googling that term). This also binds both of those classes to each other and prevents them from being able to change if/when their responsibilities change. 

For example: 

 - class A requires both class B and C in order to function and is therefore dependent on both of those classes
 - class B is also dependent on class C and already has that dependency being injected into it
 - class A indirectly obtains class C from class B (so class A is **only** injecting class B) instead of requiring it directly
 - a new business requirement comes along (like GDPR) which forces changes to existing domain logic. 
 - what if the logic in class B that requires class C needs to be extracted into a new class (class D)? Class B would no longer have an instance of class C available to pass along to class A. Now class A needs to be refactored in order for it to obtain an instance of class C which it can no longer retrieve from class B.
 - or what if it is determined that class C needs to be split into multiple subclasses (like D & E) that extend class C, and class B's requirement changes but class A's does not? So class B now requires class D (subclass of C), and class A now requires class E (also a subclass of C). This issue might be harder to spot because Class A can still obtain an object from class B that appears to be an instance of class C, but is actually now an instance of class D, however the data and/or functionality that class A requires has now been moved into class E. 

So what this means is that class B is either forced to maintain dependencies that it no longer requires (bloat), or class A is forced to change solely because of changes required in class B (the domino effect).

Going back to the simple example of "unit testing" as a reason for not hiding dependencies, and we can see that service location obviously interferes with the ability to inject mocks into a class for the purpose of testing it. If a class is obtaining its dependencies from another class, then any injected mock now has to also inject a mock of the second class. This of course is not impossible but should immediately highlight how these classes have become tangled up together. It's even worse if that service location or injection is happening via static method calls because now the class providing the dependency can not even be substituted for a mock. This is a common problem in EE right now because of our past usage of singletons.

So bottom line: Injecting a class's dependencies directly allows them to be controlled externally at runtime, which can then be done via a DI container, or a unit testcase setup method, etc. and gives classes the freedom to change when required and also to not have changes forced upon them when not required. This greatly enhances the solidity of code because it better adheres to the Single Responsibility Principle.

File: core/domain/entities/editor/blocks/widgets/EventAttendees.php (hunk 3 above)
Comment (tn3rb) [reply core/domain/entities/editor/blocks/widgets/EventAttendees.php]: ya that's probably a better way to do this...
I was trying to be clever and do everything via the parent class instead of having to add constructors and declare dependencies for all of the individual blocks, but in light of my big comment above regarding dependencies, this doesn't look so clever does it?

**Comment by tn3rb** [comment]
Ya I think:

```
{PROJECT_NAME/EPIC_NAME}|BUG|FET/[{issue-num}/]{concise-description}
```

is a better / more concise schema for what I was suggesting, although I have a couple of suggestions/clarifications:

A) I don't know if we want to ever have `{issue-num}/` as optional for anything other than the absolute base branch for a project or epic. 

ie: we **never** want branch names like:

 - FET/some-stuff-i-done
 - BUG/fix

and even the base branches for Projects or Epics can just be the name (or abbreviation) for the Project or Epic:

```
{PROJECT_NAME|EPIC_NAME}
```

B) Do we want `PROJECT_NAME` **and `EPIC_NAME` separated by `/` or do we want that to be an `OR` ? ie: are we **always** going to use Epics within Projects? And how does that affect master branch? Can we use Epics on master without a corresponding Project?

**Comment by nerrad** [comment]
> A) I don't know if we want to ever have {issue-num}/ as optional for anything other than the absolute base branch for a project or epic.

That would be true if we weren't using github.  But if we enforce the _always_ use pull request mantra in github, issue num is not as important because you have information about the branch in the pull request.   Another reason why I suggest its optional is because if one does a pull request only (i.e. there's no corresponding issue, which _will_ happen) we have a chicken/egg situation where you won't know the number of the pull request until its created.  So the switch to github makes this less of a "hard" rule as it had to be in the past.  However, I do agree if there IS an accompanying **issue** for the work, its useful to have the issue number in the branch name, I just don't think it _can_ be required.

>  Do we want PROJECT_NAME **and EPIC_NAME separated by / or do we want that to be an OR ? ie: are we always going to use Epics within Projects? And how does that affect master branch? Can we use Epics on master without a corresponding Project?

Oh, I think you misunderstood I was just using those as a OR condition mainly with the mindset that not every "project" might also be an "epic" (for instance we don't have a "Gutenberg" epic yet, but we do have a Gutenberg project.  I'm not sure we should have a hard rule on this as I imagine in _most_ cases we'll either have a project or an epic, but not both.  In the cases where we do, we could probably just do `PROJECT/EPIC` like you suggest.  But as I said, in _most_ cases I think it will end up being `PROJECT` _or_ `EPIC`.

**Comment by tn3rb** [comment]
> ut if we enforce the always use pull request mantra in github, issue num is not as important because you have information about the branch in the pull request. Another reason why I suggest its optional is because if one does a pull request only (i.e. there's no corresponding issue, which will happen) we have a chicken/egg situation where you won't know the number of the pull request until its created. 

**GREAT** points. Had not considered that but totally agree.

> Oh, I think you misunderstood I was just using those as a OR condition mainly with the mindset that not every "project" might also be an "epic" 

oh ya I see now that you also suggested that as an OR, I had looked at your full schema suggestion first:
```
{PROJECT_NAME/EPIC_NAME}|BUG|FET/[{issue-num}/]{concise-description}
```
and saw the `/` instead of `|` and then didn't notice the change when you wrote:
> Base branches for a project should simply be {PROJECT_NAME|EPIC_NAME}/base

> I'm not sure we should have a hard rule on this as I imagine in most cases we'll either have a project or an epic, but not both. In the cases where we do, we could probably just do PROJECT/EPIC like you suggest. But as I said, in most cases I think it will end up being PROJECT or EPIC.

totally agree

KK so I'm going to go ahead and do the following then:

 - rename that Project I started to simply be "Gutenberg" 
 - add all of the related Gutenberg issues and PRs to the Gutenberg Project
 - rename the `FET/Gutenberg/11400/block-manager` branch to `Gutenberg/master` (using master instead of base to keep with git standards)
 - create issues for any existing Gutenberg  related tickets in codebase
 - create branches  `Gutenberg/{issue}/{description}` branches where applicable

Final thoughts?
We should also get @mnelson4's opinion on this since I like to do things as democratically as possible

**Comment by nerrad** [comment]
@tn3rb there's been some moving around of things in this pull and based on your last comment I'm not sure exactly what I'm reviewing?  Am I reviewing the current state of code in _this_ branch?

File: core/domain/services/admin/ExitModal.php (hunk 4 above)
Comment (tn3rb) [reply core/domain/services/admin/ExitModal.php]: removed

File: core/services/collections/Collection.php (hunk 8 above)
Comment (nerrad) [reply core/services/collections/Collection.php]: > yaaa... I could go through the trouble of creating another issue and branch for every little thing added to core, but that just adds extra stuff on to everyone's plates, and since this is not a high priority need, and this class rarely changes, I don't mind it being in this branch.

Just to be clear that's not what I'm saying here (that we have to create another issue and branch for every little thing added to core).  Put another way, what I mean is if the change we are making is something that is not directly related to this branch and touches a part of the codebase that might get changed elsewhere in another branch, it's probably _safer_ to do it as an atomic commit in a branch off of master and get it merged _back_ to master right away.  I don't think that adds extra load to us and actually potentially _helps_ prevent future issues because it reduces the chance of merge conflicts.  Also, the more atomic the commit, the quicker it gets through code review and the easier it is to verify via a quick test.

If you don't think this particular file will get modified in any other branch or that you could use your change elsewhere, then sure just leave it here.

File: core/domain/entities/editor/EditorBlockInterface.php [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+     * @param string $post_type
+     * @return boolean
+     */
+    public function appliesToPostType($post_type);
+
+
+    /**
+     * @return array
+     */
+    public function getEditorContainer();
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply core/domain/entities/editor/EditorBlockInterface.php]: This was something I added based on code you wrote for the initial Gutenberg Integration Experiment plugin. 
I actually don't know what this code is for or what it is supposed to do because it wasn't documented in the original version for obvious reasons.
I think it pertains more to integrating our CPT editors with Gutenberg which we haven't gotten into fully of course, so this is very likely to change in the future.

File: core/domain/entities/editor/EditorBlock.php [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+    }
+
+
+    /**
+     * @return array
+     */
+    public function getEditorContainer()
+    {
+        return array(
+            $this->namespacedEditorBlockType(),
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply core/domain/entities/editor/EditorBlock.php]: see next comment

File: core/domain/entities/editor/EditorBlock.php [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+
+    /**
+     * Registers the Editor Block with WP core;
+     * Returns the registered block type on success, or false on failure.
+     *
+     * @return WP_Block_Type|false
+     */
+    public function registerBlock()
+    {
+        $context ='core';
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply core/domain/entities/editor/EditorBlock.php]: yup updated to get script handles from new asset register class

File: core/domain/DomainInterface.php [hunk 21]
```diff
@@ -50,6 +50,18 @@ public function pluginUrl();
...
+    /**
+     * @return string
+     */
+    public function sourceAssetsPath();
+
+
+    /**
+     * @return string
+     */
+    public function sourceAssetsUrl();
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply core/domain/DomainInterface.php]: removed

File: core/domain/DomainBase.php [hunk 22]
```diff
@@ -125,6 +132,24 @@ public function versionValueObject()
     }
 
 
+    /**
+     * @return string
+     */
+    public function sourceAssetsPath()
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply core/domain/DomainBase.php]: removed

File: core/domain/values/session/SessionLifespan.php (hunk 5 above)
Comment (tn3rb) [reply core/domain/values/session/SessionLifespan.php]: branch just needed updating. is correct now

File: core/services/assets/Registry.php (hunk 6 above)
Comment (tn3rb) [reply core/services/assets/Registry.php]: branch just needed updating. is correct now

File: src/assets/src/blocks/query-panel/index.js (hunk 9 above)
Comment (tn3rb) [reply src/assets/src/blocks/query-panel/index.js]: branch just needed updating. is correct now

**Comment by tn3rb** [comment]
> Am I reviewing the current state of code in this branch?
yes plz. 
Everything except `assets/src/blocks/widgets/event-attendees.js` which will be fixed as part of #316 

> there's been some moving around of things in this pull
The classname changes were at your request as well as many of the other changes
I *think* I've addressed all of your comments. 
My apologies again for this PR being a bit of a mess.

**Comment by nerrad** [comment]
> I pushed some other changes as well but touched issue #315 instead o f this PR, so plz also see the extra commits at the bottom of : #315

It looks like the commits touching #315 are actually in the `Gutenberg/event-attendees-block` branch and not the branch this pull request is for.  FYI, a _pull-request_ is directly attached to the branch it is referencing.  You don't have to touch issues for any commits added to a branch the pull request is for. 

I think it might be better at this point if we close this pull request and open a new one for the new branch and you can reference this pull request after creation.  Then I can just do the review on the new branch.  Right now I think its just too confusing to have code across two branches.

**Comment by tn3rb** [comment]
replaced by https://github.com/eventespresso/event-espresso-core/pull/340
